README

    Quax - A neat desktop zooming tool
  ======================================
  
  
  1) What is Quax?

    Quax is a  desktop zooming tool intended especially for web developers, but
  is very handy and friendly for newbies. There are few other tools for this
  task: XMag, KMag and KZoom. All of them are full featured applications 
  and take some time to figure how to use them. Quax is *very* easy to use.

    The only thing you need to know is to right click on Quax surface to get
  the menu.

    Web developers and graphics designer need fast access so they use shortcuts.
  Read the manual or discover them for yourself in the popup menu.
  For the manual run:

    man quax

  If you use KDE, then press Alt+F2, type "man:quax" and fire <ENTER> key.


  2) How I'll install Quax?

    Please keep in mind that you must:
      1) have Qt 4.4+ libraries installed
      2) have Qt development binaries and headers installed

    Execute:

     qmake
     make
     make install (as root)


  3) How to obtain Quax?
	
    Go to:  http://gitorious.org/quax/quax
    The old homepage was on:  http://www.ro.kde.org/quax
  

  4) Where I can report bugs and whishes?
  
    To the authors. See the AUTHORS file.


ChangeLog
2009-08-25  eros 1.2 release
	* porting: ported the app to Qt 4 (tested with 4.5+)

2003-11-20  claudiuc 1.0-1 release

2003-11-19  claudiuc
	* spec: updated SuSE and Mandrake spec to include the handbook

2003-11-17  claudiuc
	* app: corrected URL to new Quax documentation
	* spec: updated RedHat spec to include the handbook

2003-11-15  claudiuc
	* doc: finished Docbook user documentation (Quax manual)
	* app: minor labeling changes for menus 

2003-11-15  claudiuc
	* doc: finished Doxygen documentation; started Docbook user doc
	* app: now color tooltip will be repositioned when will
	get out for screen; made some optimizations, removed unused
	variables and others are local

2003-06-17  claudiuc
	* spec: support for RedHat 9.0 (script + spec)

2003-06-16  claudiuc
	* spec: support for Mandrake 9.x (script + spec)
	* doc: install instructions for Mandrake 9.x; started to
	document Quax; more Doxygen setup; updated manual page
	* app: removed qDebug()'s

2003-06-14  claudiuc 0.9-3 release

2003-06-14  claudiuc
	* win: quax.mak for buildind Quax under WIN32 completed;
	added README file for Windows users

2003-06-13  claudiuc
	* app: more #ifdef's for Windows version of hand_cursor
	and QT2/3
	* win: clean-win32.bat for cleaning up after build, improved
	quax.mak for buildind Quax under WIN32 (almost finished)


2003-06-10  claudiuc
	* app: implemented mouse wheel Quax zoom and rotation
	* win: #ifdef's and fixes to compile Quax on Qt 2.3.0 GPL
	on WIN32. It works, but "handy" should be corrected. Added
	MSVC project files and make file

2003-06-08  claudiuc 0.9-2 release

2003-06-08  claudiuc
	* doc: updated TODO and corrected README file;
	improved INSTALL instructions
	* spec: improved SPEC file and fixed the bug with
	directories permissions screewing up  
	* app: removed two qDebug(), fixed a little .destkop file,
	removed *.pro files 

2003-06-01  claudiuc
	* color picker: implemented color tooltip and color 
	picker menu. Now remain to catch undefined cases and 
	implement Ctrl+C for color
	* icon: changed Quax application icon
	* menu: now popup menu doesn't show anymore outside 
	the screen
	* interaction: left click when color tip is visible and 
	color will be copied in clipboard; pointer cursor 
	track Quax state: hand cursor, cross cursor and sizeall
	cursor
	* help: help menu work now using kfmclient from KDE to
	display Quax man page
  
2003-05-29  claudiuc
	* install: Anatoly Kamchatnov <rev@male.ru> discovered
	that PREFIX have no default value when compiling from 
	sources
  
2003-05-26  claudiuc
	* kde: added quax icons: 16x16, 32x32 and 48x48 and
	moved the *.desktop file location
	* install: SuSE spec file
	* website: Anatoly Kamchatnov <rev@male.ru> and
	Jim Richardson <warlock@eskimo.com> pointed that
	I messed up links on "Download" section
  
2003-05-25  claudiuc 0.9-1 release
  
2003-05-24  claudiuc
	* install: RPM spec file + script to build RPM
	* doc: improved docs a little
	* doc: accelerators now work ('this' is the correct
	parent for popupmenu's and not 'menu')

2003-05-24  claudiuc
	* install: made qmake based compiling system
	* docs: added GNU required docs: AUTHORS, ChangeLog etc.
	
2003-xx-xx  claudiuc
	* app: Quax in actual functionality
	
2001-xx-xx  claudiuc
	* app: implemented a simple quax



src/quax.h
//
//
// C++ Interface: h
//
// Description: Quax declarations
//
//
// Authors:  <claudiuc@kde.org>, (C) 2001, 2002, 2003
//           <enrico.ros@gmail.com>, (C) 2009
//
// Copyright: See COPYING file that comes with this distribution
//
//

#ifndef __quax_h__
#define __quax_h__

#include <QBitmap>
#include <QEvent>
#include <QIcon>
#include <QImage>
#include <QMenu>
#include <QPixmap>
#include <QTextBrowser>
#include <QWidget>

#define ZOOM_SCALE_MIN 2
#define ZOOM_SCALE_MAX 8

#define SHAPE_WIDTH 150
#define SHAPE_HEIGHT 150
#define SHAPE_INNER_X 7
#define SHAPE_INNER_Y 6
#define SHAPE_INNER_WIDTH 136
#define SHAPE_INNER_HEIGHT 136

#define UPDATE 40

/**
 * Magnifing widget for desktop
 * 
 * @short Magnifing widget for desktop
 * @author Claudiu Costin <claudiuc@kde.org>
 * @version $Id: quax.h,v 1.9 2003/11/15 20:15:15 claudiuc Exp $
 */
class Quax : public QWidget
{
    Q_OBJECT    
    public:
        /**
         * Contruct Quax GUI and initialize variables to sane values.
         * It build popup menu, set the Quax widget pixmap mask, build custom
         * cursors and start grab timer.
         */
        Quax();
        /**
         * Destroy Quax instance
         */
        ~Quax();
    
    protected:
        /**
         * Show color tooltip when mouse enter Quax area.
         */
        void enterEvent(QEvent *e);
        /**
         * Hide color tooltip if shown.
         */
        void leaveEvent(QEvent *e);
        /**
         * If right button was pressed, show popup menu. If color tooltip is already
         * showed then "take a shot" of current pixel under cursor.
         */
        void mousePressEvent(QMouseEvent *e);
        /**
         * Executed when user release move the mouse. If the color tip is enabled,
         * then displayed color is updated as well the tooltip position.
         * If is @p m_inDrag is true, then Quax position is updated (dragged)
         * to that of mouse position.
         */
        void mouseMoveEvent(QMouseEvent *e);
        /**
         * Executed when user release one mouse button.
         */
        void mouseReleaseEvent(QMouseEvent *e);
        /**
         * Get the modifier keys status and display color tooltip if Shift and Ctrl
         * are both pressed down. If cursor keys are pressed, then adjust Quax
         * position by one pixel or by ten pixels if Shift is down.
         */
        void keyPressEvent(QKeyEvent *e);
        /**
         * Hide color tooltip if one of Shift and Ctrl key modifiers is released.
         */
        void keyReleaseEvent(QKeyEvent *e);
        /**
         * Take the grabbed @p pix_zoom pixmap and scale it. Then it's painted
         * on Quax following an circle region for clipping.
         */
        void paintEvent(QPaintEvent *e);
        /**
         * Grab a screen region at every 40 miliseconds executing @ref grab().
         */
        void timerEvent(QTimerEvent *e);
        /**
         * If mouse wheel is dragged, then zoom scale in or out. When Ctrl key
         * modifier is pressed, Quax is rotated left or right
         */
        void wheelEvent(QWheelEvent *e);

    private:
        QString m_colorStringDecimal,   ///< the current color under mouse as decimals comma seperated
                m_colorStringHexaLower, ///< the current color under mouse as web RGB with lower hexadecimals
                m_colorStringHexaUpper; ///< the current color under mouse as web RGB with upper hexadecimals
        QMenu *m_menu,      ///< The main menu of Quax
              *m_menuZoom,  ///< Menu for zoom levels and zoom in and zoom out items
              *m_menuLook,  ///< The "Look at" menu
              *m_menuColor; ///< Menu for coying textual representation into clipboard
        QActionGroup *m_zoomGroup,
                     *m_rotGroup,
                     *m_colorGroup;
        QTextBrowser *m_colorTip; ///< The tool tip for displaying current color under mouse
        QPixmap m_zoomPixmap,  ///< The grabbed image zoomend and clipped
                m_colorPixmap; ///< The icon used in color menu item and color tooltip
        bool m_inDrag,         ///< if true, the user is dragging the Quax
             m_colorTipEnabled;///< if true, the color tooltip is displayed
        int m_zoomLevel, ///< current zoom level
            m_lookAt;    ///< current quadrant direction where Quax "look at"
        QPoint m_dragOffset;   ///< the offset between Quax position and new mouse position

        /**
         * It grabs the desktop area, scale it, clip it with circle region,
         * and then paint on Quax area.
         */
        void grab();
        /**
         * Show color tooltip and move the QLabel when mouse position change. Color
         * is updated runnning @ref grabForPixel(). If tooltip will be displayed
         * clipped by screen margins, then it will be repositioned before moving.
         */
        void genColorTip();
        /**
         * Grab exactly one pixel under mouse cursor to show it in color tooltip.
         */
        void grabForPixel();

    protected Q_SLOTS:
        /**
         * Display manual page or open help center according with underling desktop environment.
         */
        void slotHelp();
        /**
         * Open Quax "about" messagebox with author list and some other
         * usefull information (copyright, version).
         */
        void slotAbout();
        /**
         * Open Trolltech standard "about" messagebox.
         */
        void slotAboutQt();
        /**
         * Scale up grabed desktop area if not at maximum zoom value.
         */
        void slotZoomIn();
        /**
         * Scale down grabed desktop area if not at minimum zoom value.
         */
        void slotZoomOut();
        /**
         * Zoom grabed desktop area to the @p pos level.
         * @param pos represent position into zoom menu, as well the zoom level
         */
        void slotZoomTo();
        /**
         * Move the Quax point of desktop area grab to the North-West direction.
         * It make use of @ref Quax::rotate
         */
        void slotRotate();
        /**
         * Rotate Quax point of desktop area grab to the left (counter-clockwise).
         */
        void slotRotateLeft();
        /**
         * Rotate Quax point of desktop area grab to the right (clockwise).
         */
        void slotRotateRight();
        /**
         * Rotate Quax point of desktop area to the @p pos quadrant.
         * @param pos quadrant number (0..3)
         */
        void rotate(int pos);
        /**
         * Copy the last color under mouse which was saved on @p colorStringDecimal,
         * @p colorStringHexaLower and @p colorStringHexaUpper into system
         * clipboard. For the menu item which handle "Ctrl+C" accelerator, it call
         * @ref grabForPixel() prior to cliboard operation.
         */
        void slotColorToClipboard();
        /**
         * Build the menu left pixmap from the grabed color under mouse. This is slot is
         * called from QPopupMenu::aboutToShow signal.
         */
        void slotUpdateColorMenu();
};

#endif


src/main.cpp
//
//
// C++ Implementation: cpp
//
// Description: main application
//
//
// Authors:  <claudiuc@kde.org>, (C) 2001, 2002, 2003
//           <enrico.ros@gmail.com>, (C) 2009
//
// Copyright: See COPYING file that comes with this distribution
//
//

#include <QApplication>
#include <QLocale>
#include <QTranslator>

#include "quax.h"

int main(int argc, char ** argv)
{
    QApplication app(argc, argv);
    app.setApplicationName("Quax");
    app.setApplicationVersion("1.2");

    // install translators (qt + this application)
    QString locale = QLocale::system().name();
    QTranslator qTranslator;
    qTranslator.load(QString("qt_") + locale, ".");
    app.installTranslator(&qTranslator);
    QTranslator appTranslator;
    appTranslator.load(QString("quax_") + locale, ".");
    app.installTranslator(&appTranslator);

    Quax quax;
    quax.setWindowTitle(app.applicationName());
    quax.show();
    return app.exec();
}


makerpmrh90.sh
#!/bin/sh

export QTDIR=/usr/lib/qt-3.1

# get the current version & release
version=`cat  VERSION | cut -d- -f 1`
release=`cat  VERSION | cut -d- -f 2`

# make sure there's no objects and binaries arround
make clean
rm -f src/quax

# reconfigure things just to make sure rpmspec is good
./configure

# make a new snapshot
mkdir -p ../quax-$version-$release
cp -r . ../quax-$version-$release/

# jump a step behind
cd ..

# make archive from it and then delete local copy
tar cfz quax-$version-$release.tar.gz quax-$version-$release
rm -r quax-$version-$release

# build the RPM
cp -f  quax/quaxrh90.spec /usr/src/redhat/SPECS
cp -f  quax-$version-$release.tar.gz /usr/src/redhat/SOURCES
rpmbuild -ba /usr/src/redhat/SPECS/quaxrh90.spec
rm -f  /usr/src/redhat/SPECS/quaxrh90.spec
rm -f  /usr/src/redhat/SOURCES/quax-$version-$release.tar.gz

# back to work...
cd quax



src/quax.cpp
//
//
// C++ Implementation: cpp
//
// Description: Quax application implementation
//
//
// Authors:  <claudiuc@kde.org>, (C) 2001, 2002, 2003
//           <enrico.ros@gmail.com>, (C) 2009
//
// Copyright: See COPYING file that comes with this distribution
//
//

#include "quax.h"

#include "mag.xpm"
#include "mag-alpha.xpm"
#include "exit.xpm"
#include "viewmag.xpm"
#include "viewmagin.xpm"
#include "viewmagout.xpm"
#include "look.xpm"
#include "left.xpm"
#include "right.xpm"
#include "icon.xpm"
#include "help.xpm"

#include <QApplication>
#include <QClipboard>
#include <QColor>
#include <QCursor>
#include <QDesktopServices>
#include <QDesktopWidget>
#include <QKeyEvent>
#include <QMenu>
#include <QMessageBox>
#include <QMouseEvent>
#include <QPaintEvent>
#include <QPainter>
#include <QPixmap>
#include <QProcess>
#include <QToolTip>
#include <QUrl>
#include <QWheelEvent>


Quax::Quax()
  : QWidget(0, Qt::WindowStaysOnTopHint | Qt::CustomizeWindowHint | Qt::FramelessWindowHint)
  , m_colorStringDecimal("255, 255, 255")
  , m_colorStringHexaLower("#ffffff")
  , m_colorStringHexaUpper("#FFFFFF")
  , m_inDrag(false)
  , m_colorTipEnabled(false)
  , m_zoomLevel(ZOOM_SCALE_MIN)
  , m_lookAt(0)
{
    // customize looks
    //setAttribute(Qt::WA_NoSystemBackground);
    //setAttribute(Qt::WA_TranslucentBackground, true);
    setFixedSize(150, 150);
    setCursor(Qt::PointingHandCursor);
    setWindowIcon(QPixmap(icon_xpm));
    rotate(0);
    
    // build RMB popup menu
    m_menuZoom = new QMenu(tr("Zoom"), this);
    m_menuZoom->setIcon(QPixmap(viewmag_xpm));
    m_zoomGroup = new QActionGroup(this);
    for (int i = ZOOM_SCALE_MIN; i <= ZOOM_SCALE_MAX; i++) {
        QAction * zoomAction = m_menuZoom->addAction(tr("%1:1").arg(i), this, SLOT(slotZoomTo()));
        zoomAction->setCheckable(true);
        zoomAction->setProperty("level", i);
        zoomAction->setActionGroup(m_zoomGroup);
    }
    m_menuZoom->addSeparator();
    m_menuZoom->addAction(QPixmap(viewmagin_xpm), tr("Zoom In"), this, SLOT(slotZoomIn()), QKeySequence(tr("+","Zoom In")));
    m_menuZoom->addAction(QPixmap(viewmagout_xpm), tr("Zoom Out"), this, SLOT(slotZoomOut()), QKeySequence(tr("-","Zoom Out")));

    m_menuLook = new QMenu(tr("Look At"), this);
    m_menuLook->setIcon(QPixmap(look_xpm));
    m_rotGroup = new QActionGroup(this);
    QAction * r0 = m_menuLook->addAction(tr("North-West"), this, SLOT(slotRotate()), QKeySequence(tr("U", "North-West")));
    r0->setProperty("rotation", 0);
    r0->setActionGroup(m_rotGroup);
    QAction * r1 = m_menuLook->addAction(tr("South-West"), this, SLOT(slotRotate()), QKeySequence(tr("J", "South-West")));
    r1->setProperty("rotation", 1);
    r1->setActionGroup(m_rotGroup);
    QAction * r2 = m_menuLook->addAction(tr("North-East"), this, SLOT(slotRotate()), QKeySequence(tr("I", "North-East")));
    r2->setProperty("rotation", 2);
    r2->setActionGroup(m_rotGroup);
    QAction * r3 = m_menuLook->addAction(tr("South-Eash"), this, SLOT(slotRotate()), QKeySequence(tr("K", "South-East")));
    r3->setProperty("rotation", 3);
    r3->setActionGroup(m_rotGroup);
    m_menuLook->addSeparator();
    m_menuLook->addAction(QPixmap(left_xpm), tr("Rotate Left"), this, SLOT(slotRotateLeft()), QKeySequence(tr("L", "Rotate Left")));
    m_menuLook->addAction(QPixmap(right_xpm),tr("Rotate Right"), this, SLOT(slotRotateRight()), QKeySequence(tr("R", "Rotate Right")));

    m_menuColor = new QMenu(tr("Copy Color"), this);
    m_colorGroup = new QActionGroup(this);
    QAction * c0 = m_menuColor->addAction(m_colorStringDecimal, this, SLOT(slotColorToClipboard()));
    c0->setActionGroup(m_colorGroup);
    c0->setProperty("id", 0);
    QAction * c1 = m_menuColor->addAction(m_colorStringHexaLower, this, SLOT(slotColorToClipboard()));
    c1->setActionGroup(m_colorGroup);
    c1->setProperty("id", 1);
    QAction * c2 = m_menuColor->addAction(m_colorStringHexaUpper, this, SLOT(slotColorToClipboard()));
    c2->setActionGroup(m_colorGroup);
    c2->setProperty("id", 2);

    m_menu = new QMenu(this);
    m_menu->addMenu(m_menuZoom);
    m_menu->addMenu(m_menuLook);
    m_menu->addMenu(m_menuColor);
    m_menu->addAction(QPixmap(help_xpm), tr("Help"), this, SLOT(slotHelp()), QKeySequence(tr("H", "Help")));
    m_menu->addSeparator();
    m_menu->addAction(tr("About Quax"), this, SLOT(slotAbout()));
    m_menu->addAction(tr("About Qt"), this, SLOT(slotAboutQt()));
    m_menu->addSeparator();
    m_menu->addAction(QPixmap(exit_xpm),tr("Quit"), qApp, SLOT(quit()), QKeySequence(tr("Q", "Quit")));
    connect(m_menu, SIGNAL(aboutToShow()), this, SLOT(slotUpdateColorMenu()));

    m_zoomGroup->actions().first()->setChecked(true);
    m_rotGroup->actions().at(1)->setChecked(true);
    
    // set the color tooltip
    m_colorTip = new QTextBrowser;
    m_colorTip->setWindowFlags(Qt::WindowStaysOnTopHint | Qt::CustomizeWindowHint | Qt::FramelessWindowHint | Qt::Tool | Qt::X11BypassWindowManagerHint);
    m_colorTip->setAlignment(Qt::AlignLeft | Qt::AlignTop);
    m_colorTip->setFrameStyle(QFrame::Plain | QFrame::Box);
    m_colorTip->resize(120, 28);
    
    setMouseTracking(true);
    startTimer(UPDATE);
}

Quax::~Quax()
{
    delete m_colorTip;
}

void Quax::enterEvent(QEvent *)
{
    if (m_colorTipEnabled && m_colorTip->isHidden())
        genColorTip();
}

void Quax::leaveEvent(QEvent *)
{
    if (m_colorTipEnabled)
        m_colorTip->hide();
}

void Quax::paintEvent(QPaintEvent *)
{
    if (m_zoomPixmap.isNull())
        return;

    QPainter p(this);
    p.setClipping(true);
    p.setClipRegion(QRegion(SHAPE_INNER_X,SHAPE_INNER_Y,SHAPE_INNER_WIDTH,SHAPE_INNER_HEIGHT,QRegion::Ellipse));

    int x,y;
    if (m_zoomPixmap.width() > SHAPE_INNER_WIDTH)
        x=(m_zoomPixmap.width() - SHAPE_INNER_WIDTH)/2;
    else
        x=(SHAPE_INNER_WIDTH - m_zoomPixmap.width())/2;
    if (m_zoomPixmap.height() > SHAPE_INNER_HEIGHT)
        y=(m_zoomPixmap.height() - SHAPE_INNER_HEIGHT)/2;
    else
        y=(SHAPE_INNER_HEIGHT - m_zoomPixmap.height())/2;
    x += SHAPE_INNER_X;
    y += SHAPE_INNER_Y;
    p.drawPixmap(x, y, m_zoomPixmap, 0, 0, SHAPE_INNER_WIDTH, SHAPE_INNER_HEIGHT);
}

void Quax::timerEvent(QTimerEvent *)
{
    grab();
}

void Quax::wheelEvent(QWheelEvent *e)
{
    if (e->modifiers() & Qt::ControlModifier) {
        if (e->delta() > 0)
            slotRotateLeft();
        else
            slotRotateRight();
    } else {
        if (e->delta() > 0)
            slotZoomIn();
        else
            slotZoomOut();
    }
    e->accept();
    if (m_colorTipEnabled && !m_colorTip->isHidden())
        genColorTip();
}

void Quax::grab()
{
    int x, y, w, h;
    int grabX, grabY, grabW, grabH, ofsX, ofsY;

    w = SHAPE_INNER_WIDTH / m_zoomLevel;
    h = SHAPE_INNER_HEIGHT / m_zoomLevel;
    switch (m_lookAt) {
        case 0: // South-West
        default:
            x = this->x()-w;
            y = this->y()+height();
            break;
        case 1: // North-West
            x = this->x()-w;
            y = this->y()-h;
            break;
        case 2: // North-East
            x = this->x()+width();
            y = this->y()-h;
            break;
        case 3: // South-East
            x = this->x()+width();
            y = this->y()+height();
            break;
    }

    // calculate grab surface to avoid garbage
    if (x < 0) {
        grabX = 0;
        grabW = x+w;
        ofsX  = -x;
    } else if (x+w > qApp->desktop()->width()-1) {
        grabX = x;
        grabW = qApp->desktop()->width()-1-x;
        ofsX  = 0;
    } else {
        grabX = x;
        grabW = w;
        ofsX  = 0;
    }
    if (y < 0) {
        grabY = 0;
        grabH = y+h;
        ofsY  = -y;
    } else if (y+h > qApp->desktop()->height()-1) {
        grabY = y;
        grabH = qApp->desktop()->height()-1-y;
        ofsY  = 0;
    } else {
        grabY = y;
        grabH = h;
        ofsY  = 0;
    }

    QPixmap grabPix(w, h);
    grabPix.fill(Qt::blue);
    if (grabW > 0 && grabH > 0) {
        QPixmap pix_temp = QPixmap::grabWindow(QApplication::desktop()->winId(), grabX, grabY, grabW, grabH);
        QPainter p(&grabPix);
        p.drawPixmap(ofsX,ofsY,pix_temp);
    }

    m_zoomPixmap = grabPix.transformed(QTransform::fromScale((double)m_zoomLevel,(double)m_zoomLevel));
    update();
}

void Quax::grabForPixel()
{
    // get the color under cursor
    QPixmap grabPix = QPixmap::grabWindow(QApplication::desktop()->winId(), QCursor::pos().x() - 1, QCursor::pos().y() - 1, 1, 1);
    QRgb rgb_color = grabPix.toImage().pixel(0, 0);
    m_colorStringDecimal = QString::number(qRed(rgb_color)) + ", " +
                           QString::number(qGreen(rgb_color)) + ", " +
                           QString::number(qBlue(rgb_color));
    m_colorStringHexaLower.sprintf("#%02x%02x%02x",qRed(rgb_color),qGreen(rgb_color),qBlue(rgb_color));
    m_colorStringHexaUpper = m_colorStringHexaLower.toUpper();

    // make a little pixmap with grabbed color
    m_colorPixmap = grabPix.scaled(14, 14);
}

void Quax::genColorTip()
{
    grabForPixel();
    m_colorTip->document()->addResource(QTextDocument::ImageResource, QUrl("qrc:/pixel"), m_colorPixmap);
    m_colorTip->document()->setHtml("<html>&nbsp;<img height=\"14\" width=\"14\" src=\"qrc:/pixel\">&nbsp;<tt>" + m_colorStringHexaUpper + "</tt></html>");

    int x_pos, y_pos;
    int x_ofs = 7, y_ofs = 5;
    int x_margin = 2, y_margin = 2;
    //check if tooltip get outside screen area
    QRect  r = m_colorTip->rect();
    QPoint c = QCursor::pos();
    if (c.x()+r.width()+x_ofs+x_margin > qApp->desktop()->width())
        x_pos = c.x()-r.width()-x_ofs;
    else
        x_pos = c.x()+x_ofs;
    if (c.y()-r.height()-y_ofs-y_margin < 0)
        y_pos = c.y()+y_ofs;
    else
        y_pos = c.y()-r.height()-y_ofs;
    m_colorTip->move(x_pos,y_pos);
    m_colorTip->show();
}

void Quax::mousePressEvent(QMouseEvent *e)
{
    switch (e->button()) {
        case Qt::LeftButton:
            if (m_colorTipEnabled) {
                slotColorToClipboard();
                setCursor(Qt::SizeAllCursor);
            } else {
                grabMouse(Qt::PointingHandCursor);
                m_inDrag = true;
                m_dragOffset = QCursor::pos() - QPoint(geometry().left(),geometry().top());
            }
            break;
        case Qt::RightButton:
            m_menu->popup(QCursor::pos());
            break;
        default:
            break;
    }
}

void Quax::mouseMoveEvent(QMouseEvent *)
{
    if (m_colorTipEnabled)
        genColorTip();
    if (m_inDrag && this == mouseGrabber())
        move(QCursor::pos() - m_dragOffset);
}

void Quax::mouseReleaseEvent(QMouseEvent *e)
{
    switch (e->button()) {
        case Qt::LeftButton:
            if (m_colorTipEnabled)
                setCursor(Qt::CrossCursor);
            else {
                releaseMouse();
                m_inDrag = false;
            }
            break;
        default:
            break;
    }
}

void Quax::keyPressEvent(QKeyEvent *e)
{
    if (!m_colorTipEnabled && e->modifiers() & Qt::ShiftModifier && e->modifiers() & Qt::ControlModifier) {
        e->ignore();
        m_colorTipEnabled = true;
        setCursor(Qt::CrossCursor);
        genColorTip();
        return;
    }

    QPoint dist(0, 0);
    switch (e->key()) {
        case Qt::Key_Left:
            dist.setX(-1);
            break;
        case Qt::Key_Right:
            dist.setX(1);
            break;
        case Qt::Key_Up:
            dist.setY(-1);
            break;
        case Qt::Key_Down:
            dist.setY(1);
            break;
        case Qt::Key_Equal:
        case Qt::Key_Plus:
            slotZoomIn();
            break;
        case Qt::Key_R:
            slotRotateRight();
            break;
        case Qt::Key_Underscore:
        case Qt::Key_Minus:
            slotZoomOut();
            break;
        default:
            QWidget::keyPressEvent(e);
            return;
    }
    if (e->modifiers() & Qt::ShiftModifier)
        dist *= 10;
    move(pos() + dist);
}

void Quax::keyReleaseEvent(QKeyEvent *e)
{
    if (m_colorTipEnabled) {
        if (e->key() == Qt::Key_Control || e->key() == Qt::Key_Shift) {
            m_colorTipEnabled = false;
            m_colorTip->hide();
            setCursor(Qt::PointingHandCursor);
        }
    }
    QWidget::keyReleaseEvent(e);
}

void Quax::slotHelp()
{
    if (QDesktopServices::openUrl(QUrl("help:quax")))
        return;
    QMessageBox::warning(this, tr("Help Error"),
                          tr("<qml>There was an loading the help application."
                             "Please check your KDE installation. You may try to open an <b>xterm</b> "
                             "window and run <tt>man quax</tt> from there.</qml>"),
                          QMessageBox::Ok);
}

void Quax::slotAbout()
{
    QMessageBox::about(this, tr("About Quax %1").arg(qApp->applicationVersion()), tr(
            "<p><b>Quax</b> is a little magnifing tool for X. Quax homepage is "
            "at <tt>&lt;http://www.ro.kde.org/quax/&gt;</tt>."
            "</p><p>This is Quax version %1."
            "</p><p>This program is released under GNU General Public License."
            "<dl>"
            "<dt><nobr>Claudiu Costin <tt>&lt;claudiuc@kde.org&gt;</tt></nobr></dt>"
            "<dd>coding, documentation, packaging</dd>"
            "<dt><nobr>Enrico Ros <tt>&lt;enrico.ros@gmail.com&gt;</tt></nobr></dt>"
            "<dd>porting to Qt4 and cleanups</dd>"
            "<dt><nobr>Vlad Oancea <tt>&lt;xtrim@fx.ro&gt;</tt></nobr></dt>"
            "<dd>Quax skin design</dd>"
            "</dl></p>").arg(qApp->applicationVersion()));
}

void Quax::slotAboutQt()
{
    QMessageBox::aboutQt(this, tr("About Qt"));
}

void Quax::slotZoomIn()
{
    if (m_zoomLevel < ZOOM_SCALE_MAX) {
        m_zoomLevel++;
        foreach (QAction * action, m_zoomGroup->actions())
            if (action->property("level").toInt() == m_zoomLevel)
                action->setChecked(true);
    }
}

void Quax::slotZoomOut()
{
    if (m_zoomLevel > ZOOM_SCALE_MIN) {
        m_zoomLevel--;
        foreach (QAction * action, m_zoomGroup->actions())
            if (action->property("level").toInt() == m_zoomLevel)
                action->setChecked(true);
    }
}

void Quax::slotZoomTo()
{
    QAction * action = static_cast<QAction *>(sender());
    m_zoomLevel = action->property("level").toInt();
}

void Quax::slotRotate()
{
    QAction * action = static_cast<QAction *>(sender());
    m_lookAt = action->property("rotation").toInt();
    rotate(m_lookAt);
}

void Quax::slotRotateLeft()
{
    m_lookAt = !m_lookAt ? 3 : m_lookAt - 1;
    m_rotGroup->actions().at(m_lookAt)->setChecked(true);
    rotate(m_lookAt);
}

void Quax::slotRotateRight()
{
    m_lookAt = m_lookAt == 3 ? 0 : m_lookAt + 1;
    m_rotGroup->actions().at(m_lookAt)->setChecked(true);
    rotate(m_lookAt);
}

void Quax::rotate(int pos)
{
    QTransform rot;
    rot.rotate(pos * 90.0);
    QPalette pal;
    pal.setBrush(QPalette::Window, QPixmap(mag_xpm).transformed(rot));
    setPalette(pal);
    setMask(QPixmap(mag_alpha_xpm).transformed(rot).mask());
}

void Quax::slotColorToClipboard()
{
    QAction * action = dynamic_cast<QAction *>(sender());
    int id = action ? action->property("id").toInt() : 2;
    switch (id) {
        case 0:
            qApp->clipboard()->setText(m_colorStringDecimal);
            break;
        case 1:
            // this get called when user press Ctrl+C
            //
            // I must call grabForPixel because it's posible that
            // this slot get executed even user doesn't right click
            // on Quax, so slotUpdateColorMenu is not yet called and
            // variables may contain old or invalid color
            grabForPixel();
            qApp->clipboard()->setText(m_colorStringHexaLower);
            break;
        case 2:
            qApp->clipboard()->setText(m_colorStringHexaUpper);
            break;
    }
}

void Quax::slotUpdateColorMenu()
{
    grabForPixel();
    m_menuColor->setIcon(QIcon(m_colorPixmap));
    m_colorGroup->actions().at(0)->setText(m_colorStringDecimal);
    m_colorGroup->actions().at(1)->setText(m_colorStringHexaLower);
    m_colorGroup->actions().at(2)->setText(m_colorStringHexaUpper);
}
